File: 270. Closest Binary Search Tree Value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def closestValue(self, root: TreeNode, target: float) -> int:
        res = float('inf')
        store = None
        while root:
            if root.val > target:
                if res > root.val - target:
                    store = root.val
                    res = root.val - target
                root = root.left
            else:
                if res > target - root.val:
                    store = root.val
                    res = target - root.val
                root = root.right
        return store

File: 270. Closest Binary Search Tree Value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int closestValue(TreeNode root, double target) {
        int clmax = Integer.MIN_VALUE, clmin = Integer.MAX_VALUE;
        while (root != null) {
            if ((double)root.val > target) {
                clmax = root.val;
                root = root.left;
            } else if ((double)root.val < target) {
                clmin = root.val;
                root = root.right;
            } else {
                return root.val;
            }
        }
        if (clmax == Integer.MIN_VALUE) {
            return clmin;
        } else if (clmin == Integer.MAX_VALUE) {
            return clmax;
        } else {
            return (target-clmin > clmax-target) ? clmax : clmin;
        }
    }
}